### ISSUE TITLE
Application crash from date to unix conversion

### ISSUE THREAD

**Issue by allanbreyes** [jangernert/FeedReader#692]
Version: `master`

Stack trace:
```
An error occured (SIGSEGV) in ../src/DataBaseWriteAccess.vala, line 504 in 'feed_reader_data_base_write_articles'
   The reason is likely a null reference being used.

    #1  <unknown>                                             in 'g_date_time_to_unix'
        at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
 *  #2  ../src/DataBaseWriteAccess.vala            line  504 in 'feed_reader_data_base_write_articles'
        at /home/allan/Workspace/FeedReader/src/DataBaseWriteAccess.vala:504
    #3  ../src/Backend/FeedServerInterface.vala    line  111 in 'feed_reader_feed_server_interface_getArticles'
        at /home/allan/Workspace/FeedReader/src/Backend/FeedServerInterface.vala:111
    #4  ../src/Backend/FeedServer.vala             line 1013 in 'feed_reader_feed_server_getArticles'
        at /home/allan/Workspace/FeedReader/src/Backend/FeedServer.vala:1013
    #5  ../src/Backend/FeedServer.vala             line  254 in 'feed_reader_feed_server_syncContent'
        at /home/allan/Workspace/FeedReader/src/Backend/FeedServer.vala:254
```

Appears to be an [LoD](https://en.wikipedia.org/wiki/Law_of_Demeter) error on [L504](https://github.com/jangernert/FeedReader/blob/7de9c3238fe7c5cb39c8fe4c5917ef88b3c9c79e/src/DataBaseWriteAccess.vala#L504)... but I'm stumped as to why it's possibly throwing since it's guaranteed set/defined here: https://github.com/jangernert/FeedReader/blob/7de9c3238fe7c5cb39c8fe4c5917ef88b3c9c79e/src/Model/Article.vala#L80. :man_shrugging:

**Comment by brendanlong** [comment]
It looks like `SetDate()` doesn't have that check :\

**Comment by allanbreyes** [comment]
Uhh, still flapping with Vala, but yikes... looks like [DateTime.now_utc()](https://valadoc.org/glib-2.0/GLib.DateTime.DateTime.now_utc.html) can return `null`. Wat?

**Comment by allanbreyes** [comment]
Alright... I'm not sure what's going on with this. Might be something on my local machine, after reading `DateTime.now`:

> This function will always succeed unless the system clock is set to truly insane values (or unless GLib is still being used after the year 9999 ).

Not sure what's up with my system clock... seems like it's sane and in the right year... going to "tap out" on this one, as it's a bit out of my league. Any help resolving this issue would be most appreciated. Thanks!

**Comment by allanbreyes** [comment]
Oh, @brendanlong, to clarify: I changed `SetDate()` to:

```vala
	public void SetDate(GLib.DateTime date)
	{
		m_date = date != null ? date : new DateTime.now_utc();
	}
```

... but still saw the same error. Again, my Vala-fu is weak, but I'm convinced it's from `getDate()` somehow returning `null`.

Might also be something wrong with my build environment... I see **lots** of things getting nulled out:

```
[about 1000+ lines like this before]
** (feedreader:18960): CRITICAL **: 07:38:28.921: feed_reader_data_base_read_only_article_exists: assertion 'articleID != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.921: feed_reader_article_getHTML: assertion 'self != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.921: feed_reader_article_getHTML: assertion 'self != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.921: string_replace: assertion 'self != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.921: feed_reader_article_setHTML: assertion 'self != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.921: feed_reader_article_getArticleID: assertion 'self != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.921: feed_reader_data_base_read_only_article_exists: assertion 'articleID != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.921: feed_reader_article_getHTML: assertion 'self != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.921: feed_reader_article_getHTML: assertion 'self != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.921: string_replace: assertion 'self != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.921: feed_reader_article_setHTML: assertion 'self != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.921: feed_reader_article_getArticleID: assertion 'self != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.921: feed_reader_data_base_read_only_article_exists: assertion 'articleID != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.921: feed_reader_article_getHTML: assertion 'self != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.921: feed_reader_article_getHTML: assertion 'self != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.921: string_replace: assertion 'self != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.921: feed_reader_article_setHTML: assertion 'self != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.921: feed_reader_article_getArticleID: assertion 'self != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.921: feed_reader_data_base_read_only_article_exists: assertion 'articleID != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.921: feed_reader_article_getHTML: assertion 'self != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.921: feed_reader_article_getHTML: assertion 'self != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.921: string_replace: assertion 'self != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.921: feed_reader_article_setHTML: assertion 'self != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.922: feed_reader_article_getDate: assertion 'self != NULL' failed

(feedreader:18960): GLib-CRITICAL **: 07:38:28.922: g_date_time_difference: assertion 'end != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.922: feed_reader_article_getDate: assertion 'self != NULL' failed

(feedreader:18960): GLib-CRITICAL **: 07:38:28.922: g_date_time_difference: assertion 'end != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.922: feed_reader_article_getArticleID: assertion 'self != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.922: feed_reader_article_getFeedID: assertion 'self != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.922: feed_reader_article_getURL: assertion 'self != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.922: feed_reader_article_getUnread: assertion 'self != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.922: feed_reader_article_getMarked: assertion 'self != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.922: feed_reader_article_getTitle: assertion 'self != NULL' failed

(feedreader:18960): feedreader-WARNING **: 07:38:28.922: Utils.UTF8fix: string is NULL

** (feedreader:18960): CRITICAL **: 07:38:28.922: feed_reader_article_getHTML: assertion 'self != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.922: feed_reader_article_getPreview: assertion 'self != NULL' failed

(feedreader:18960): feedreader-WARNING **: 07:38:28.922: Utils.UTF8fix: string is NULL

** (feedreader:18960): CRITICAL **: 07:38:28.922: feed_reader_article_getAuthor: assertion 'self != NULL' failed

** (feedreader:18960): CRITICAL **: 07:38:28.922: feed_reader_article_getDate: assertion 'self != NULL' failed

(feedreader:18960): feedreader-WARNING **: 07:38:28.922: getDate: (null)

** (feedreader:18960): CRITICAL **: 07:38:28.922: feed_reader_article_getDate: assertion 'self != NULL' failed
An error occured (SIGSEGV) in ../src/DataBaseWriteAccess.vala, line 507 in 'feed_reader_data_base_write_articles'
   The reason is likely a null reference being used.

    #1  <unknown>                                             in 'g_date_time_to_unix'
        at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
 *  #2  ../src/DataBaseWriteAccess.vala            line  507 in 'feed_reader_data_base_write_articles'
        at /home/allan/Workspace/FeedReader/src/DataBaseWriteAccess.vala:507
    #3  ../src/Backend/FeedServerInterface.vala    line  111 in 'feed_reader_feed_server_interface_getArticles'
        at /home/allan/Workspace/FeedReader/src/Backend/FeedServerInterface.vala:111
    #4  ../src/Backend/FeedServer.vala             line 1013 in 'feed_reader_feed_server_getArticles'
        at /home/allan/Workspace/FeedReader/src/Backend/FeedServer.vala:1013
    #5  ../src/Backend/FeedServer.vala             line  254 in 'feed_reader_feed_server_syncContent'
        at /home/allan/Workspace/FeedReader/src/Backend/FeedServer.vala:254
```

**Comment by brendanlong** [comment]
Hmmm maybe something is wrong with the article parser and it's generating a bad `Article`. What backend are you using?

**Comment by brendanlong** [comment]
Oh something in TT-RSS :\ I don't really know how to test that backend. I bet their API is sending us something weird and our parser is getting confused. The logs might have useful information about what's going on near the beginning of the output.

**Comment by jangernert** [comment]
Yeah, definitely something out of whack there. You can do a `message.printResponse();` after it was sent in the ttrssAPI.

https://github.com/jangernert/FeedReader/blob/master/plugins/backend/ttrss/ttrssAPI.vala#L573
Just need to figure out which call fails. Hopefully the logs will be of help there.

**Comment by brendanlong** [comment]
I saw similar problems with the TT-RSS plugin which were fixed by a52474a7fea023abc6ff9b873db6a0067f091254. Can you try v2.8.0 and see if it works for you now?

I'm closing the bugs I think are fixed by this release but please re-open it if you're still seeing the same problem!
